### PR TITLE
Add initial support for RSockets

### PR DESCRIPTION
**Pull request by artembilan**
* Add `spring-integration-rsocket` module and respective dependencies
* Implement `RSocketOutboundGateway` based on the Spring Messaging
`RSocketRequester`.
This component supports dynamic RSocket properties via expressions
against request message.
to handle `Publisher` for requests, it must be present in the request
message `payload` instead of `FluxMessageChannel` upstream, since the
last one just flattens events to be handled in the `MessageHandler` one
by one.
The result `Mono` is subscribed downstream in the `FluxMessageChannel`
or directly by the `AbstractReplyProducingMessageHandler`.
If result is a `Flux` it is just wrapped into the `Mono` to be processed
downstream by end-user code.
The point is that these request/replies are volatile and live in the
particular context meanwhile a `FluxMessageChannel` is long living
publisher in the application context boundaries.
* The `RSocketOutboundGatewayIntegrationTests` is an adapted copy of
`RSocketClientToServerIntegrationTests` from Spring Messaging
* Add `doOnError()` into the `Flux` created in the
`AbstractMessageProducingHandler` for `Publisher` replies

### CODE REVIEW

**Comment by artembilan** [comment]
@garyrussell ,

You can merge if it is OK: other features and Docs can be added in the upcoming PRs.
Or let me know and I'll continue in this branch with the rest of stuff

Thanks

File: build.gradle [hunk 1]
```diff
@@ -596,6 +598,18 @@ project('spring-integration-rmi') {
 	}
 }
 
+project('spring-integration-rsocket') {
+	description = 'Spring Integration RSockets Support'
```
Comment (garyrussell) [inline build.gradle]: It appears this should be singular `RSocket`.

File: spring-integration-rsocket/src/main/java/org/springframework/integration/rsocket/outbound/RSocketOutboundGateway.java [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+	public void setExpectedResponseTypeExpression(Expression expectedResponseTypeExpression) {
+		this.expectedResponseTypeExpression = expectedResponseTypeExpression;
+	}
+
+
+	@Override
+	protected void doInit() {
+		super.doInit();
+		RSocketRequester.Builder builder = RSocketRequester.builder();
+		if (this.factoryConfigurer != null) {
```
Comment (garyrussell) [inline spring-integration-rsocket/src/main/java/org/springframework/integration/rsocket/outbound/RSocketOutboundGateway.java]: It's probably cleaner to default to a no-op configurers and assert not null in setters.

File: spring-integration-rsocket/src/main/java/org/springframework/integration/rsocket/outbound/RSocketOutboundGateway.java [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+/**
+ * An Outbound Messaging Gateway for RSocket client requests.
+ *
+ * @author Artem Bilan
+ *
+ * @since 5.2
+ *
+ * @see RSocketRequester
+ */
+public class RSocketOutboundGateway extends AbstractReplyProducingMessageHandler {
```
Comment (garyrussell) [inline spring-integration-rsocket/src/main/java/org/springframework/integration/rsocket/outbound/RSocketOutboundGateway.java]: Shouldn't this be `RSocketMessageHandler` since it supports `fireAndForget` (adapter semantics)?

File: spring-integration-rsocket/src/test/java/org/springframework/integration/rsocket/outbound/RSocketOutboundGatewayIntegrationTests.java [hunk 4]
```diff
@@ -0,0 +1,437 @@
...
+/**
+ * @author Artem Bilan
+ *
+ * @since 5.2
+ */
+@SpringJUnitConfig
+@DirtiesContext
+public class RSocketOutboundGatewayIntegrationTests {
+
+	private static final int PORT = SocketUtils.findAvailableTcpPort();
```
Comment (garyrussell) [inline spring-integration-rsocket/src/test/java/org/springframework/integration/rsocket/outbound/RSocketOutboundGatewayIntegrationTests.java]: Is there no support for port 0 so the OS allocates the port?

**Review by garyrussell — CHANGES_REQUESTED** [review]
Good work! Just a few issues.

File: spring-integration-rsocket/src/main/java/org/springframework/integration/rsocket/outbound/RSocketOutboundGateway.java (hunk 3 above)
Comment (artembilan) [reply spring-integration-rsocket/src/main/java/org/springframework/integration/rsocket/outbound/RSocketOutboundGateway.java]: I thought about that, too, but there is already a `RSocketMessageHandler` in Spring Messaging and its point is about a *server* side, so we don't have a choice and choose a different name.

Re. `fireAndForget` (a-la one-way): it is really doesn't fit to reactive model. We always need to return something (in this case a `Mono<Void>`) to let end-user to decide when to subscribe to it and, essentially, perform a request.
So, even if it is empty, it still going to be a reply from this gateway. The `FluxMessageChannel` downstream will subscribe to that `Mono` or `async` mode will do an eager subscription otherwise.

In summary: we always return something from here, even if it is just about `fireAndForget`

Comment (garyrussell) [reply spring-integration-rsocket/src/main/java/org/springframework/integration/rsocket/outbound/RSocketOutboundGateway.java]: ok

File: spring-integration-rsocket/src/test/java/org/springframework/integration/rsocket/outbound/RSocketOutboundGatewayIntegrationTests.java (hunk 4 above)
Comment (artembilan) [reply spring-integration-rsocket/src/test/java/org/springframework/integration/rsocket/outbound/RSocketOutboundGatewayIntegrationTests.java]: I just can't swallow code complexity in Netty, so it isn't clear if it is there.
@rstoyanchev just uses a static `7000` port on the matter 😢 

/CC @smaldini

Comment (garyrussell) [reply spring-integration-rsocket/src/test/java/org/springframework/integration/rsocket/outbound/RSocketOutboundGatewayIntegrationTests.java]: It's just that we went through a process of converting all our TCP module tests to use port 0 so the OS chooses the port. These "findAvailablePort" schemes, while they work most of the time, we occasionally saw failures (BindException) on CI builds. A hard-wired port # doesn't sound good either.

Comment (artembilan) [reply spring-integration-rsocket/src/test/java/org/springframework/integration/rsocket/outbound/RSocketOutboundGatewayIntegrationTests.java]: OK. I have figured out how to use `0` for port.

See latest commit.

Thanks

File: spring-integration-rsocket/src/test/java/org/springframework/integration/rsocket/outbound/RSocketOutboundGatewayIntegrationTests.java [hunk 5]
```diff
@@ -74,7 +74,7 @@
 @DirtiesContext
 public class RSocketOutboundGatewayIntegrationTests {
 
-	private static final int PORT = SocketUtils.findAvailableTcpPort();
+	private static int PORT;
```
Comment (garyrussell) [inline spring-integration-rsocket/src/test/java/org/springframework/integration/rsocket/outbound/RSocketOutboundGatewayIntegrationTests.java]: This should no longer be static.

Comment (artembilan) [reply spring-integration-rsocket/src/test/java/org/springframework/integration/rsocket/outbound/RSocketOutboundGatewayIntegrationTests.java]: ??? It is initialized from the `static @BeforeAll` and it won't be available in the `ClientConfig` below which has to be a `static` as well.

Comment (garyrussell) [reply spring-integration-rsocket/src/test/java/org/springframework/integration/rsocket/outbound/RSocketOutboundGatewayIntegrationTests.java]: Sorry; I meant it should no longer be upper case.
